### PR TITLE
dupeguru: init at 4.0.4

### PR DESCRIPTION
**Pull request by novoxd**
###### Motivation for this change
Add dupeGuru programm https://dupeguru.voltaicideas.net .
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by novoxudonoser** [comment]
Just found that I need to rewrite source fetching, from fetchGit to fetchFromGitHub with submodules.

**Comment by novoxudonoser** [comment]
Please review, I don't know how long it takes to get a review in nixpkgs, but there are now auto-suggested reviewers. Maybe @doronbehar can help.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (doronbehar) [inline pkgs/applications/misc/dupeguru/default.nix]: Why not `python.withPackages`?

File: pkgs/applications/misc/dupeguru/default.nix
Comment (doronbehar) [inline pkgs/applications/misc/dupeguru/default.nix]: There's no need to import both `python37` and `python37Packages`. Use here just `python` or `pythonPackages` and when you `callPackage`, use `python = `python3;`. Also, use just the `3` suffix in `callPackage` so it will use the latest and greatest Python version as time goes on.

**Review by doronbehar — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/dupeguru/default.nix
Comment (novoxudonoser) [reply pkgs/applications/misc/dupeguru/default.nix]: @doronbehar 
>and when you callPackage

`callPackage` in `pkgs/top-level/all-packages.nix` ?

**Comment by doronbehar** [comment]
Yes.

**Review by doronbehar — APPROVED** [review]

**Comment by novoxudonoser** [comment]
@doronbehar I don't know  - but what next? Do I need to wait for this pr to be merged? Or I need to mention someone ho can add his review and merge? (I am quite new to nixpkgs)

**Comment by doronbehar** [comment]
It's OK, welcome aboard the Caos :smiley_cat:. You _can_ mention maybe `@jonringer` or `@worldofpeace` if days pass and nothing happens but you better get used to this - PRs can take infinitely long time to be merged..

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: It is recommended to avoid explicit phases if you can:

```suggestion
  buildFlags = [
    "i18n"
    "modules"
    "qt/dg_rc.py"
    "pyc"
  ];
```

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: There should not be a need to call this manually, unless you disable it.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]:
```suggestion
    installFlags = [
      "PREFIX=${placeholder ''out''}"
      "--assume-old=all"
      "--assume-old=pyc"
    ];
```

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]:
```suggestion
python3.pkgs.buildPythonApplication rec {
```

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: Better just use python3.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: The python dependencies can go to `pythonPath` attribute directly in `buildPythonApplication`.

Plus `format = "other"` to call `make` instead of `setup.py`.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/dupeguru/default.nix]: Though I actually recommend the following to avoid double wrapping.

```
  dontWrapQtApps = true;

  preFixup = ''
    makeWrapperArgs+=("''${gappsWrapperArgs[@]}")
  '';
```

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: With `buildPythonApplication` `python3.pkgs.setuptools` can go directly to `nativeBuildInputs`, no need to add build Python.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: Maintainers? Platforms?

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/dupeguru/default.nix
Comment (novoxudonoser) [reply pkgs/applications/misc/dupeguru/default.nix]: Is there any instruction on how to find any fitable?  Or any list of nixpkgs maintainers?

File: pkgs/applications/misc/dupeguru/default.nix
Comment (novoxudonoser) [reply pkgs/applications/misc/dupeguru/default.nix]: @jtojnar what about https://github.com/NixOS/nixpkgs/pull/75054#discussion_r355661631 ? Is it acceptable advice?

File: pkgs/applications/misc/dupeguru/default.nix
Comment (novoxudonoser) [reply pkgs/applications/misc/dupeguru/default.nix]: @jtobin what about `chmod +x $out/share/dupeguru/run.py` as I guess the python file won't get executable.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/dupeguru/default.nix]: `platforms = platforms.linux;` if you tested it on Linux, `unix` if you can also test it on MacOS.

Generally, when you add a new package, you volunteer to become its maintainer (see `maintainers` entry in https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes). Or you can try finding someone interested in maintaining this in the community.

**Comment by novoxudonoser** [comment]
> It's OK, welcome aboard the Caos . You _can_ mention maybe `@jonringer` or `@worldofpeace` if days pass and nothing happens but you better get used to this - PRs can take infinitely long time to be merged..

@doronbehar thanks for your review and help.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/dupeguru/default.nix]: The advice about using only one of `pythonXxx` or `pythonXxxPackages` is sound. Otherwise, when user overrides one of them and forgets to override the other one, you would get multiple Pythons.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/dupeguru/default.nix]: Is that really necessary? I would not expect scripts in `$out/share` to be executed directly. But if it is necessary, you can add it to `postInstall` and add a comment explaining how is it getting executed.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]:
```
  dontWrapQtApps = true;

  preFixup = ''
    makeWrapperArgs+=("''${qtWrapperArgs[@]}")
  '';
```

This should be sufficient to prevent double wrapping.

**Comment by novoxudonoser** [comment]
@jtojnar I have some problems after trying to make things like in your review.

I get error:
```
✦ ➜ /nix/store/5l9ckal7javjd04qa4n2afsmlal9naj5-dupeguru-4.0.4/bin/dupeguru
Traceback (most recent call last):
  File "/nix/store/5l9ckal7javjd04qa4n2afsmlal9naj5-dupeguru-4.0.4/bin/dupeguru", line 12, in <module>
    from PyQt5.QtCore import QCoreApplication, QSettings
ModuleNotFoundError: No module named 'PyQt5'
```
It looks like required packages are not in the python interpreter what wraps `out/bin/dupguru` . I had the same error on start but here I don't know what to do or how to analyze the problem. Can you give any advice about figuring out why it fails?

But as I know I made default.nix correct, all by instructions. Can you please review changes, maybe I am mistaken somewhere? The problem can be in dupguru itself its Makefile is a little "broken"

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: This does not appear to be necessary.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: This does not appear to be necessary. File is already executable.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/dupeguru/default.nix]: Okay, the issue is that the GUI executable is in in `share`, the thing in `bin` is just a symlink. And Python infrastructure only sets up things in `bin`:

https://github.com/NixOS/nixpkgs/blob/8d5e90bdb32c7f8708a5b75eaaca00551e2f1d6b/pkgs/development/interpreters/python/mk-python-derivation.nix#L149
https://github.com/NixOS/nixpkgs/blob/8d5e90bdb32c7f8708a5b75eaaca00551e2f1d6b/pkgs/development/interpreters/python/wrap.sh#L4

Adding

```
postFixup = ''
  wrapPythonProgramsIn "$out/share/dupeguru/run.py" "$out $pythonPath"
'';

should fix that.

**Comment by doronbehar** [comment]
@novoxudonoser Try to find some python apps in the collection which use qt and wrap it and you should be able to get an idea as for how to fix your issue.

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/dupeguru/default.nix]: Hmm, there is a bug in the Python wrapper, this line

https://github.com/NixOS/nixpkgs/blob/8d5e90bdb32c7f8708a5b75eaaca00551e2f1d6b/pkgs/development/interpreters/python/wrap.sh#L84

needs to be 

to work around it, the `preFixup` section should go like

```
    # TODO: A bug in python wrapper
    # see 
    makeWrapperArgs="''${qtWrapperArgs[@]}"
```

Comment (jtojnar) [reply pkgs/applications/misc/dupeguru/default.nix]: Sorry, should have been `"$out/share/dupeguru"` not `"$out/share/dupeguru/run.py"`.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: There are some trailing spaces in the file.

**Comment by novoxudonoser** [comment]
@jtojnar made all as you said. The problem is still there. Seems that we are missing something. The most interesting thin that in https://github.com/NixOS/nixpkgs/pull/75054/commits/d409fe39cee63f84846c75c3a5761fb5e446fa30 all worked, I guess that the problem is - that using `buildPythonApplication` for some bugs or wrong configuration same actions like in https://github.com/NixOS/nixpkgs/pull/75054/commits/d409fe39cee63f84846c75c3a5761fb5e446fa3 are not reproduced.

**Comment by jtojnar** [comment]
Sorry, should have been "$out/share/dupeguru" not "$out/share/dupeguru/run.py".

**Comment by novoxudonoser** [comment]
> @novoxudonoser Try to find some python apps in the collection which use qt and wrap it and you should be able to get an idea as for how to fix your issue.

While the building of the first version of derivation I looked to many examples. And I picked `stdenv.mkDerivation` cause the make file in dupguru is a little "broken" (seems like symlink to python file breaks nixpkgs standart `buildPythonApplication` process ). Rith now I don't have enough knowledge to understand what breaks here. I think more deep research in nix build process and python section can help, and I plan to do it as part of my education process.

**Comment by novoxudonoser** [comment]
@jtojnar finally it worked! Thanks a lot for your help, how I said upper - I don't have enough knowledge to make this derivation correct.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: This is typically placed between `pname`+`version` and `src` to make it immediately obvious.

Comment (novoxudonoser) [reply pkgs/applications/misc/dupeguru/default.nix]: fixed

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]:
```suggestion
    # Executable in $out/bin is a symlink to $out/share/dupeguru/run.py
    # wrapPythonPrograms hook does not handle it automatically.
    wrapPythonProgramsIn "$out/share/dupeguru" "$out $pythonPath"
```

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: We tend to quote URLs these days, see https://github.com/NixOS/rfcs/pull/45

**Comment by jtojnar** [comment]
> @jtojnar finally it worked! Thanks a lot for your help, how I said upper - I don't have enough knowledge to make this derivation correct on my own.

Unfortunately, bash is such a terrible language that lot of these things are more complicated than they should be.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]:
```suggestion
    description = "GUI tool to find duplicate files in a system";
```

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: Sorry if I were unclear, I meant below `version` and above `src`.

```
  pname = "dupeguru";
  version = "4.0.4";

  format = "other";

  src = fetchFromGitHub {
```

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: Could you also add comment why is assume-old necessary.

**Comment by jtojnar** [comment]
Looking at the `makeFile`:

It appears that the `buildFlags` are already part of the default `make` target (`make all`):
https://github.com/arsenetar/dupeguru/blob/2ea02bd7b50ce12ba489bc90717c4b578641deaf/Makefile#L47

It also suggests that we pass `NO_VENV=1` to `make`:

https://github.com/arsenetar/dupeguru/blob/2ea02bd7b50ce12ba489bc90717c4b578641deaf/Makefile#L22-L24

I recommend dropping `buildFlags` and `installFlags` and just using:

```nix
  makeFlags = [
    "PREFIX=${placeholder ''out''}"
    "NO_VENV=1"
  ];
```

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]:
```suggestion
  # Avoid double wrapping Python programs.
  dontWrapQtApps = true;
```

**Comment by novoxudonoser** [comment]
> I recommend dropping `buildFlags` and `installFlags` and just using:
> 
> ```nix
>   makeFlags = [
>     "PREFIX=${placeholder ''out''}"
>     "NO_VENV=1"
>   ];
> ```

Nice! I did not even think about this.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/dupeguru/default.nix]: We could probably do

```
  checkInputs = with python3Packages; [
    pytest
    pytest-monkeyplus
  ];

  pytestFlagsArray = [ "core" "hscommon" ];
```

but that would require packaging `pytest-monkeyplus`.

If you do not feel like packaging it, please add a comment explaining that we are missing some test dependencies.

**Comment by jtojnar** [comment]
Also please split adding a maintainer into a separate commit a squash all dupeguru fixes.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (novoxudonoser) [reply pkgs/applications/misc/dupeguru/default.nix]: I think that it will be a nice task for another pr. I don't feel myself ready to do it right now.

Comment (jtojnar) [reply pkgs/applications/misc/dupeguru/default.nix]: Could you add it as a comment then?

```suggestion
  # TODO: package pytest-monkeyplus for running tests
  # https://github.com/NixOS/nixpkgs/pull/75054/files#r357690123
  doCheck = false;
```

Comment (novoxudonoser) [reply pkgs/applications/misc/dupeguru/default.nix]: @jtojnar fixed.

**Review by jtojnar — CHANGES_REQUESTED** [review]
Looks good to me, great job.

**Review by jtojnar — APPROVED** [review]
Looks good to me, great job.

**Comment by novoxudonoser** [comment]
@jtojnar - what next?  Is there any need to mentions somebody who can make an additional review and merge pr?

**Comment by jtojnar** [comment]
I set it to merge once CI succeeds.

**Comment by novoxudonoser** [comment]
@jtojnar thanks you a lot again, you made my day, +1 maintainer to nixpkgs.

File: pkgs/applications/misc/dupeguru/default.nix
Comment (lovesegfault) [reply pkgs/applications/misc/dupeguru/default.nix]: Fixing this

Comment (lovesegfault) [reply pkgs/applications/misc/dupeguru/default.nix]: I take that back, testing this change on my laptop is impossible, @worldofpeace tag!

**Pull request merged**
